src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/RazorSemanticTokensEndpoint.cs

```diff
@@ -93,13 +68,13 @@ public async Task<SemanticTokens> Handle(SemanticTokensRangeParams request, Canc
                 throw new ArgumentNullException(nameof(request));
             }
 
-            var (documentSnapshot, documentVersion) = await TryGetCodeDocumentAsync(request.TextDocument.Uri.GetAbsolutePath(), cancellationToken);
-            if (documentSnapshot is null || documentVersion is null)
+            var documentPath = request.TextDocument.Uri.GetAbsolutePath();
+            if (documentPath is null)
             {
                 return null;
             }
 
-            var edits = await _semanticTokensInfoService.GetSemanticTokensEditsAsync(documentSnapshot, request.TextDocument, documentVersion, request.PreviousResultId, cancellationToken);
+            var edits = await _semanticTokensInfoService.GetSemanticTokensEditsAsync(documentPath, request.TextDocument, request.PreviousResultId, cancellationToken);
```

Comment:
Firstly I'm a silly goose and we can get an easy win by removing `documentPath` and calculating that inside our functions since it's derived from `TextDocumentIdentifier`.

You might make an argument that `ConvertSemanticRangesToSemanticTokens` belongs in the endpoint rather than inside `DefaultRazorSemanticTokensInfoService` because it's converting the base data into a format that we're only using because of the protocol, but that gets kind of messy. Then we also have to move the _semanticTokensCache to the Endpoint as well (because it stores the completed data (it must do so  because we won't have the old document to re-calculate against on the next request)). moving THAT up means you also need the `codeDocument`, so now that's an `out` param on GetSemanticTokensAsync (and we don't have a `GetSemanticTokensEditsAsync` at all because everything that makes it distinct from GetSemanticTokensAsync moved up a level). I would call the Edit calculations business logic because they're making a determination about what the requester needs based on what they've been given in the past, and to my mind business logic belongs in the service rather than the endpoint.

Reply:
Ah so if I read what you wrote correctly the layering would be something like: Endpoint -> Semantic service -> semantic provider:

- Endpoint: Glue to handle the request
- Service: Stateful being that communicates in LSP concepts and translates various semantic tokens into edits.
- Provider: "Get me the Razor semantic tokens for said code document"

Did i interpret that correctly?

Reply:
Yes on Endpoint and Service, but I'm not sure what component you're calling a provider. Are you suggesting a third class (or renaming/repurposing of TagHelperSemanticRangeVisitor) which does the "dumb" parts (walking the tree, getting the C# response and ordering the list of the two)? If you're refering to the `ProvideSemanticTokensAsync` method I don't really see that as having enough logic to justify it's own class, it's just shelling out to the C# endpoint.

Reply:
Oo I was just trying to interpret what you said. I thought you were suggesting another class to fulfill the repsonsibility of resolving semantic tokens for just the Razor document (not C#).

I'd imagine the "shelling" out to C# could technically be its own thing but as you said it's a minor part of the puzzle. If anything I'd figure that'd be an input into the semantic token service which then aggregates the tokens into something LSP understands.

Reply:
Offline convo, no action required now but food for though!:
```
- SemanticTokenEndpoint 
	- SemanticTokenService
		-> GetSemanticTokenEditAsync 
		- SemanticTokenProvider
			-> GetSemanticTokensAsync
				Aggregate
				(
					-> ResolveSemanticTokensAsync (Razor)
					-> ResolveSemanticTokensAsync (C#)
					-> ResolveSemanticTokensAsync (HTML)
				)
```